File: web/src/types/template.ts
import type { DocumentType } from './document';

export type FieldType = 'text' | 'number' | 'date' | 'rut' | 'currency';

export interface TemplateField {
  id: string;
  name: string;
  label: string;
  field_type: FieldType;
  pattern: string;
  pattern_flags: string;
  required: boolean;
  order: number;
  validation_rules: Record<string, unknown> | null;
  post_processing: string | null;
}

export interface Template {
  id: string;
  name: string;
  code: string;
  description: string | null;
  doc_type: DocumentType;
  is_active: boolean;
  priority: number;
  classification_keywords: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface TemplateDetail extends Template {
  fields: TemplateField[];
}

export interface TemplateListResponse {
  items: Template[];
  total: number;
}

export interface TemplateFieldCreate {
  name: string;
  label: string;
  field_type: FieldType;
  pattern: string;
  pattern_flags?: string;
  required?: boolean;
  order?: number;
  validation_rules?: Record<string, unknown>;
  post_processing?: string;
}

export interface TemplateCreate {
  name: string;
  code: string;
  description?: string;
  doc_type: DocumentType;
  is_active?: boolean;
  priority?: number;
  classification_keywords?: string[];
  fields?: TemplateFieldCreate[];
}

File: api/schemas/template.py
from __future__ import annotations
from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from models.document import DocumentType, FieldType


class TemplateFieldCreate(BaseModel):
    name: str
    label: str
    field_type: FieldType = FieldType.TEXT
    pattern: str
    pattern_flags: str = "IGNORECASE"
    required: bool = False
    order: int = 0
    validation_rules: Optional[dict] = None
    post_processing: Optional[str] = None


class TemplateFieldResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    label: str
    field_type: FieldType
    pattern: str
    pattern_flags: str
    required: bool
    order: int
    validation_rules: Optional[dict] = None
    post_processing: Optional[str] = None


class TemplateCreate(BaseModel):
    name: str
    code: str
    description: Optional[str] = None
    doc_type: DocumentType
    is_active: bool = True
    priority: int = 0
    classification_keywords: Optional[List[str]] = None
    fields: List[TemplateFieldCreate] = []


class TemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    priority: Optional[int] = None
    classification_keywords: Optional[List[str]] = None


class TemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    code: str
    description: Optional[str] = None
    doc_type: DocumentType
    is_active: bool
    priority: int
    classification_keywords: Optional[List[str]] = None
    created_at: datetime
    updated_at: datetime


class TemplateDetailResponse(TemplateResponse):
    fields: List[TemplateFieldResponse] = []


class TemplateListResponse(BaseModel):
    items: List[TemplateResponse]
    total: int
